src/viewport/telemetry_safety.py

```
@@ -0,0 +1,326 @@
+def sanitize_span_attributes(attributes: Mapping[str, Any]) -> dict[str, Any]:
+    """Sanitize OpenTelemetry span attributes before export.
+
+    Auto-instrumentation can attach raw URLs, user agents, client addresses, and
+    SQL statements. This function keeps useful low-cardinality attributes and
+    fingerprints/deletes sensitive values.
+    """
+
+    return {str(key): sanitize_span_attribute(str(key), value) for key, value in attributes.items()}
+
+
+def sanitize_span_attribute(key: str, value: Any) -> Any:
```

Ruff (0.15.12)

[warning] Too many return statements (15 > 6)

(PLR0911)